lib/spack/spack/binary_distribution.py

```diff
@@ -37,23 +39,294 @@
+        # Finally, if there are configured mirrors for which we don't have a
+        # cache entry, we need to fetch and cache the indices from those
+        # mirrors.
+
+        for cached_mirror_url in self._local_index_cache:
+            cache_entry = self._local_index_cache[cached_mirror_url]
+            cached_index_hash = cache_entry['index_hash']
+            cached_index_path = cache_entry['index_path']
+            if cached_mirror_url in configured_mirror_urls:
+                # May need to fetch the index and update the local caches
+                self.fetch_and_cache_index(
+                    cached_mirror_url, expect_hash=cached_index_hash)
+            else:
+                # No longer have this mirror, cached index should be removed
+                items_to_remove.append({
+                    'url': cached_mirror_url,
+                    'cache_key': os.path.join(self._index_cache_root,
+                                              cached_index_path)
+                })
+
+        # Clean up items to be removed, identified above
+        for item in items_to_remove:
+            url = item['url']
+            cache_key = item['cache_key']
+            self._index_file_cache.remove(cache_key)
+            del self._local_index_cache[url]
+
+        # Iterate the configured mirrors now.  Any mirror urls we do not
+        # already have in our cache must be fetched, stored, and represented
+        # locally.
+        for mirror_url in configured_mirror_urls:
+            if mirror_url not in self._local_index_cache:
+                # Need to fetch the index and update the local caches
+                self.fetch_and_cache_index(mirror_url)
+
+        self._write_local_index_cache()
+
+    def fetch_and_cache_index(self, mirror_url, expect_hash=None):
+        """ Fetch a buildcache index file from a remote mirror and cache it.
+
+        If we already have a cached index from this mirror, then we first
+        check if the hash has changed, and we avoid fetching it if not.
+
+        Args:
+            mirror_url (str): Base url of mirror
+            expect_hash (str): If provided, this hash will be compared against
+                the index hash we retrieve from the mirror, to determine if we
+                need to fetch the index or not.
+        """
+        index_fetch_url = url_util.join(
+            mirror_url, _build_cache_relative_path, 'index.json')
+        hash_fetch_url = url_util.join(
+            mirror_url, _build_cache_relative_path, 'index.json.hash')
+
+        fetched_hash = None
+
+        # Fetch the hash first so we can check if we actually need to fetch
+        # the index itself.
+        try:
+            _, _, fs = web_util.read_from_url(hash_fetch_url)
+            fetched_hash = codecs.getreader('utf-8')(fs).read()
+        except (URLError, web_util.SpackWebError) as url_err:
+            tty.debug('Unable to read index hash {0}'.format(
```

Comment:
(question) I'm assuming this is an expected pathway for older binary caches which don't have an `index.json.hash` file?

Reply:
That's correct, if we don't have it on the mirror, we print a debug message and keep going.  Later on, we check whether we actually got it before trying to compare it to the hash we compute locally on the fetched index.